qgis/examples/elevation_tile_downloader/core/task.py

```diff
@@ -1,117 +1,122 @@
+                    )
+                    continue
+
+                # get the new tile if it didn't exist
+                url = QUrl(
+                    f"http://s3.amazonaws.com/elevation-tiles-prod/skadi/{dir_name}/{tile_name}.gz"
+                )
+                request = QNetworkRequest(url)
+                reply: QgsNetworkReplyContent = nam.blockingGet(request)
+                # turn the reply into a file object and write the decompressed file to disk
+                with gzip.GzipFile(
+                    fileobj=BytesIO(reply.content()), mode="rb"
+                ) as gz:
+                    with open(filepath, "wb") as f:
+                        f.write(gz.read())
+
+                self.setProgress((i / len(self.grid)) * 100)
+
             return True
         except Exception as e:
             self.exception = e
-            self.traceback = traceback.format_exc()
             return False
 
     def finished(self, result):
-        if result:
-            QgsMessageLog.logMessage(
-                message=f'Finished downloading tiles. {len(self.tiles)} tiles downloaded.',
-                level=Qgis.Info
-            )
-            if self.add:
-                self.add_to_map()
-        elif not result and self.exception:
+        """Postprocessing after the run() method."""
+        if self.isCanceled():
+            # if it was canceled by the user
             QgsMessageLog.logMessage(
-                message=f'An exception occurred during downloading: {self.exception}\n'
-                        f'{self.traceback}',
-                level=Qgis.Warning
+                message=f"Canceled download task. {len(self.tiles)} tiles downloaded.",
+                level=Qgis.Warning,
             )
-        elif not result and not self.exception:
-            QgsMessageLog.logMessage(
-                message=f'Canceled download task. {len(self.tiles)} tiles downloaded.',
-                level=Qgis.Warning
+            return
+        elif not result:
+            # if there was an error
+            QMessageBox.critical(
+                iface.mainWindow(),
+                "Tile download error",
+                f"The following error occurred:\n{self.exception.__class__.__name__}: {self.exception}",
             )
+            return
 
-    @staticmethod
-    def prepare_dir(y, dest_dir):
-        dir_name = '%s%02d' % ('S' if y < 0 else 'N', abs(y))
-        directory = Path(dest_dir, dir_name)
-        Path.mkdir(directory, exist_ok=True)
+        iface.messageBar().pushMessage(
+            "Success",
+            f"Finished downloading {len(self.tiles)} tiles.",
+            level=Qgis.Success,
+        )
 
-    def add_to_map(self):
-        project = QgsProject.instance()
+        # add the rasters to the map canvas
         for tile_path in self.tiles:
-            finfo = QFileInfo(str(tile_path))
-            base_name = finfo.baseName()
-            path = finfo.filePath()
```

Comment:
never used that one before :) I think it's more obvious for a python dev to work with the built-ins